src/Storages/MergeTree/DataPartStorageOnDisk.cpp

```diff
@@ -406,27 +406,87 @@ void DataPartStorageOnDisk::clearDirectory(
     }
 }
 
-std::string DataPartStorageOnDisk::getRelativePathForPrefix(Poco::Logger * log, const String & prefix, bool detached) const
+std::optional<String> DataPartStorageOnDisk::getRelativePathForPrefix(Poco::Logger * log, const String & prefix, bool detached, bool broken) const
 {
     String res;
 
     auto full_relative_path = fs::path(root_path);
     if (detached)
         full_relative_path /= "detached";
 
+    std::optional<String> original_checksums_content;
+    std::optional<Strings> original_files_list;
+
     for (int try_no = 0; try_no < 10; ++try_no)
     {
         res = (prefix.empty() ? "" : prefix + "_") + part_dir + (try_no ? "_try" + DB::toString(try_no) : "");
 
         if (!volume->getDisk()->exists(full_relative_path / res))
             return res;
 
+        if (broken && looksLikeBrokenDetachedPartHasTheSameContent(res, original_checksums_content, original_files_list))
+        {
+            LOG_WARNING(log, "Directory {} (to detach to) already exists, "
+                        "but its content looks similar to content of the broken part which we are going to detach. "
+                        "Assuming it was already cloned to detached, will not do it again to avoid redundant copies of broken part.", res);
+            return {};
+        }
+
         LOG_WARNING(log, "Directory {} (to detach to) already exists. Will detach to directory with '_tryN' suffix.", res);
     }
 
     return res;
 }
 
+bool DataPartStorageOnDisk::looksLikeBrokenDetachedPartHasTheSameContent(const String & detached_part_path,
```

Comment:
Now it's my favorite function name!

Reply:
Cool, my favorite is `MergeTreeData::removePartsInRangeFromWorkingSetAndGetPartsToRemoveFromZooKeeper`